src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs

```diff
@@ -20,20 +20,23 @@ public class FirstNonShouldlyMethodFinder : ITestMethodFinder
 
         public TestMethodInfo GetTestMethodInfo(StackTrace stackTrace, int startAt = 0)
         {
-            var i = startAt;
-            StackFrame callingFrame;
-            do
+            for (var i = startAt; stackTrace.GetFrame(i) is { } frame; i++)
             {
-                callingFrame = stackTrace.GetFrame(i++);
-            } while (callingFrame.GetMethod().IsShouldlyMethod() || IsCompilerGenerated(callingFrame.GetMethod()));
+                if (frame.GetMethod() is { } method && !method.IsShouldlyMethod() && !IsCompilerGenerated(method))
+                {
+                    var callingFrame = stackTrace.GetFrame(i + Offset)
+                        ?? throw new InvalidOperationException("There is no stack frame at the specified offset from the first non-Shouldly stack frame.");
 
-            callingFrame = stackTrace.GetFrame(i + Offset - 1);
-            return new TestMethodInfo(callingFrame);
+                    return new TestMethodInfo(callingFrame);
+                }
+            }
+
+            throw new InvalidOperationException("Cannot find a non-Shouldly method in the stack trace.");
```

Comment:
Recommend using the split view for these sections. If any of this is too exotic, I'm more than happy to rewrite.

Reply:
good point